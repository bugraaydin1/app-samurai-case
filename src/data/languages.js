export const languages = [
	{
		title: "English",
		code: "en",
		enterText: "Enter text",
		translation: "Translation",
		loading: "Translating...",
		noHistory: "History not found",
		history: "Translation History",
		flag: "ğŸ‡¬ğŸ‡§",
	},
	{
		title: "TÃ¼rkÃ§e",
		code: "tr",
		enterText: "Metin giriniz",
		translation: "Ã‡eviri",
		loading: "Ã‡eviri YapÄ±lÄ±yor...",
		noHistory: "GeÃ§miÅŸ arama bulunamadÄ±",
		history: "Ã‡eviri GeÃ§miÅŸi",
		flag: "ğŸ‡¹ğŸ‡·",
	},
];
