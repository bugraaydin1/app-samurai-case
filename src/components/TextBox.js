import { useState } from "react";
import LabelBox from "./LabelBox";
import Speaker from "./svgComponents/Speaker";
import History from "./svgComponents/History";
import Microphone from "./svgComponents/Microphone";
import { speakTTS } from "../utils/audio";
import { languages } from "../data/languages";
import { Language } from "../data/models/language";
import PropTypes from "prop-types";

const TextBox = ({
	type,
	loading,
	selectedLanguage,
	setTextToTranslate,
	textToTranslate,
	translatedText,
	setTranslatedText,
	onShowHistory,
}) => {
	const [recording, setRecording] = useState(false);

	const text = type === "output" ? translatedText : textToTranslate;

	const handleDelete = () => {
		setTextToTranslate("");
		setTranslatedText("");
	};

	const handleRecordToggle = () => {
		if (!recording) {
			handleSpeechRecognition();
		}
		setRecording(!recording);
	};

	const handleSpeechRecognition = () => {
		const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition;

		if (SpeechRecognition !== undefined) {
			const recognition = new SpeechRecognition();

			recognition.continuous = false;
			recognition.lang = selectedLanguage.code;
			recognition.interimResults = true;
			recognition.maxAlternatives = 1;

			recognition.start();

			recognition.onspeechend = () => {
				recognition.stop();
				setRecording(false);
			};

			recognition.onresult = (result) => {
				const resolved = result.results[0][0];
				const accuracy = Math.floor(resolved.confidence * 100);

				console.log(`${accuracy}% accurate: ${resolved.transcript}`);

				setTextToTranslate(resolved.transcript.toLowerCase());

				// for mac safari to stop rec.
				recognition.stop();
			};
		} else {
			console.warn("Speech recognition is not supported ❎");
		}
	};

	const handleTextToSpeech = () => {
		speakTTS(text, selectedLanguage.code, loading);
	};

	const getPlaceholderText = () => {
		const currentLanguage = languages.find((lang) => lang.code === selectedLanguage.code);
		return type === "input" ? currentLanguage.enterText : currentLanguage.translation;
	};

	return (
		<div className={type}>
			<LabelBox className={type} selectedLanguage={selectedLanguage.title} />
			<textarea
				className={type}
				disabled={type === "output"}
				placeholder={getPlaceholderText()}
				onChange={(e) => setTextToTranslate(e.target.value)}
				value={type === "input" ? textToTranslate : translatedText}
			/>
			{type === "input" && (
				<div className="action-buttons">
					<div className="icon-button delete" onClick={handleDelete}>
						˟
					</div>
					<div
						data-testid="microphone-icon"
						onClick={handleRecordToggle}
						className={`icon-button microphone ${recording ? "microphone-rec" : ""}`}
					>
						<Microphone />
					</div>
					<div
						data-testid="speaker-icon"
						className={"icon-button speaker"}
						onClick={handleTextToSpeech}
					>
						<Speaker />
					</div>
					<div data-testid="history-icon" onClick={onShowHistory} className={"icon-button history"}>
						<History />
					</div>
				</div>
			)}

			{type === "output" && (
				<div
					data-testid="speaker-icon"
					className={"icon-button speaker"}
					onClick={handleTextToSpeech}
				>
					<Speaker />
				</div>
			)}
		</div>
	);
};

export default TextBox;

TextBox.propTypes = {
	type: PropTypes.string.isRequired,
	loading: PropTypes.bool.isRequired,
	selectedLanguage: Language.isRequired,
	setTextToTranslate: PropTypes.func,
	textToTranslate: PropTypes.string,
	translatedText: PropTypes.string,
	setTranslatedText: PropTypes.func,
	onShowHistory: PropTypes.func,
};

TextBox.defaultProps = {
	type: "input",
	loading: false,
	selectedLanguage: languages[0],
	setTextToTranslate: () => {},
	textToTranslate: "",
	translatedText: "",
	setTranslatedText: () => {},
	onShowHistory: () => {},
};
