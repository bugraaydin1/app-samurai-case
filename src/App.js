import { useCallback, useEffect, useState } from "react";
import axios from "./config/axios";

import TextBox from "./components/TextBox";
import HistoryBox from "./components/HistoryBox";
import Arrows from "./components/svgComponents/Arrows";
import { languages } from "./data/languages";

const App = () => {
	const [inputLanguage, setInputLanguage] = useState(languages[0]);
	const [outputLanguage, setOutputLanguage] = useState(languages[1]);
	const [textToTranslate, setTextToTranslate] = useState("");
	const [translatedText, setTranslatedText] = useState("");
	const [translateLoading, setTranslateLoading] = useState(false);

	const [showHistory, setShowHistory] = useState(false);
	const [history, setHistory] = useState([]);

	const debounceTime = 800;

	const saveToHistory = useCallback(
		(textToTranslate, translate) => {
			let history = JSON.parse(localStorage.getItem(`${inputLanguage.code}`) ?? "[]");

			const isAlreadyInHistory = history.some((obj) => obj.source === textToTranslate);

			if (!isAlreadyInHistory) {
				localStorage.setItem(
					`${inputLanguage.code}`,
					JSON.stringify([...history, { source: textToTranslate, target: translate }])
				);
			}
		},
		[inputLanguage.code]
	);

	const handleStorageHistory = useCallback(() => {
		const storageHistory = JSON.parse(localStorage.getItem(`${inputLanguage.code}`) ?? "[]");
		setHistory(storageHistory);
	}, [inputLanguage.code]);

	const translate = useCallback(async () => {
		setTranslateLoading(true);

		const data = {
			q: textToTranslate,
			source: inputLanguage.code,
			target: outputLanguage.code,
		};

		const response = await axios.post("/translate", data);
		const translate = response.data.translatedText?.toLowerCase();

		setTranslatedText(translate);
		saveToHistory(textToTranslate, translate);
		handleStorageHistory();
		setTranslateLoading(false);
	}, [
		textToTranslate,
		inputLanguage.code,
		outputLanguage.code,
		handleStorageHistory,
		saveToHistory,
	]);

	useEffect(() => {
		handleStorageHistory();
	}, []);

	useEffect(() => {
		let translateTimer;
		if (textToTranslate) {
			translateTimer = setTimeout(() => {
				translate();
			}, debounceTime);
		}

		return () => {
			clearTimeout(translateTimer);
		};
	}, [textToTranslate, translate]);

	const handleToggleLanguages = () => {
		setTextToTranslate(translatedText);
		setTranslatedText(textToTranslate);
		setInputLanguage(outputLanguage);
		setOutputLanguage(inputLanguage);
	};

	return (
		<div className="app">
			<div className="flex">
				<TextBox
					type="input"
					className="input"
					loading={translateLoading}
					selectedLanguage={inputLanguage}
					setTextToTranslate={setTextToTranslate}
					textToTranslate={textToTranslate}
					setTranslatedText={setTranslatedText}
					onShowHistory={() => setShowHistory(!showHistory)}
				/>
				<div
					data-testid="arrow-container"
					className="arrow-container"
					onClick={handleToggleLanguages}
				>
					<Arrows />
				</div>
				<TextBox
					type="output"
					loading={translateLoading}
					selectedLanguage={outputLanguage}
					translatedText={!translateLoading ? translatedText : inputLanguage.loading}
				/>
			</div>

			{showHistory && (
				<HistoryBox loading={translateLoading} history={history} selectedLanguage={inputLanguage} />
			)}
		</div>
	);
};

export default App;
