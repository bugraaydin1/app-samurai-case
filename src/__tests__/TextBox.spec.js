/* eslint-disable testing-library/no-render-in-setup */
import * as React from "react";
import { screen, render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import TextBox from "../components/TextBox";
import { languages } from "../data/languages";
import App from "../App";

describe("TextBox", () => {
	const language = languages[0];

	const setTextToTranslateMock = jest.fn();
	const setTranslatedTextMock = jest.fn();
	const onShowHistoryMock = jest.fn();

	const textBoxComponent = (props) => (
		<TextBox
			selectedLanguage={language}
			setTextToTranslate={setTextToTranslateMock}
			setTranslatedTextMock={setTranslatedTextMock}
			onShowHistory={onShowHistoryMock}
			{...props}
		/>
	);

	it("renders correctly", () => {
		const textBox = renderer.create(textBoxComponent()).toJSON();

		expect(textBox).toMatchSnapshot();
	});

	it("title is correct when type is 'input'", () => {
		render(textBoxComponent({ type: "input" }));

		const input = screen.getByPlaceholderText(language.enterText);

		expect(input).toBeInTheDocument();
	});

	it("renders correct items when type is 'input'", () => {
		render(textBoxComponent({ type: "input" }));

		const icons = screen.getAllByTestId((id) => id.includes("-icon"));
		const iconClasses = icons.map((icon) => icon.className.split(" ")[1]).sort();
		const isAllIconIncluded = ["history", "microphone", "speaker"].every((ico) =>
			iconClasses.includes(ico)
		);

		expect(isAllIconIncluded).toBe(true);
	});

	it("title is correct when type is 'output'", () => {
		render(textBoxComponent({ type: "output" }));

		const input = screen.getByPlaceholderText(language.translation);

		expect(input).toBeInTheDocument();
	});

	it("renders correct items when type is 'output'", () => {
		render(textBoxComponent({ type: "output" }));

		const icon = screen.getByTestId((id) => id.includes("-icon"));
		const iconClass = icon.className.split(" ")[1];
		const isIconIncluded = iconClass === "speaker";

		expect(isIconIncluded).toBe(true);
	});
});
