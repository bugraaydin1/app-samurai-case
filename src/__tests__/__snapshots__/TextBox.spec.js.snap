// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox renders correctly 1`] = `
<div
  className="input"
>
  <div
    className="label-box"
  >
    <input
      readOnly={true}
      value="English"
    />
  </div>
  <textarea
    className="input"
    disabled={false}
    onChange={[Function]}
    placeholder="Enter text"
    value=""
  />
  <div
    className="action-buttons"
  >
    <div
      className="icon-button delete"
      onClick={[Function]}
    >
      ËŸ
    </div>
    <div
      className="icon-button microphone "
      data-testid="microphone-icon"
      onClick={[Function]}
    >
      <svg
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12,14c1.66,0,3-1.34,3-3V5c0-1.66-1.34-3-3-3S9,3.34,9,5v6C9,12.66,10.34,14,12,14z"
        />
        <path
          d="M17,11c0,2.76-2.24,5-5,5s-5-2.24-5-5H5c0,3.53,2.61,6.43,6,6.92V21h2v-3.08c3.39-0.49,6-3.39,6-6.92H17z"
        />
      </svg>
    </div>
    <div
      className="icon-button speaker"
      data-testid="speaker-icon"
      onClick={[Function]}
    >
      <svg
        height="24px"
        viewBox="0 0 24 24"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M14,3.23V5.29C16.89,6.15 19,8.83 19,12C19,15.17 16.89,17.84 14,18.7V20.77C18,19.86 21,16.28 21,12C21,7.72 18,4.14 14,3.23M16.5,12C16.5,10.23 15.5,8.71 14,7.97V16C15.5,15.29 16.5,13.76 16.5,12M3,9V15H7L12,20V4L7,9H3Z"
        />
      </svg>
    </div>
    <div
      className="icon-button history"
      data-testid="history-icon"
      onClick={[MockFunction]}
    >
      <svg
        height="25"
        viewBox="0 0 24 24"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.5,8H12V13L16.28,15.54L17,14.33L13.5,12.25V8M13,3A9,9 0 0,0 4,12H1L4.96,16.03L9,12H6A7,7 0 0,1 13,5A7,7 0 0,1 20,12A7,7 0 0,1 13,19C11.07,19 9.32,18.21 8.06,16.94L6.64,18.36C8.27,20 10.5,21 13,21A9,9 0 0,0 22,12A9,9 0 0,0 13,3"
        />
      </svg>
    </div>
  </div>
</div>
`;
